### ISSUE TITLE
Compatibility: Neovim, Don't really have a good way to explain in words....

### ISSUE THREAD

**Issue by henil** [mosaic-org/mosaic#115]
GIF to save us :smile: 

Two issues:
* Going in insert mode and pressing `Enter/Return`, the cursor goes to the beginning of the line.

* After exiting neovim, there are still some leftover parts that are not cleared properly. Not sure but I think those only happen in the area where there was a spacing issue previously. Issue #93 

![output](https://user-images.githubusercontent.com/71698300/102630787-fbb5b980-4172-11eb-99bc-e564b2452d14.gif)

**Comment by imsnif** [comment]
Thanks for the gif :) Looks like there are a bunch of issues here (eg. the 'l' appearing at the beginning of the line - or does that usually happen?).. also the line on the right (or does that also usually happen?)

Anyway, if you can open separate issues for all of these (if you have the time, of course) that would be great.
Also, could you include a debug log for them? That would really help recreate them.

**Comment by henil** [comment]
Oh sorry for the confusion, this is how my normal nvim window look:

![Screenshot from 2020-12-18 21-37-29](https://user-images.githubusercontent.com/71698300/102635459-5a7e3180-4179-11eb-8b8a-26606d0bb77a.png)

So both those '|' are not the issues.

One on the left is just to create more spacing for numbers and one on the middle is to highlight the `80` character limit. Also, some space that is seen in this screenshot on the far right is nvim not able to fit perfectly with this particular width.

**Comment by henil** [comment]
Followed the same process in this too, but this doesn't cause panic: [mosaic-3.log](https://github.com/mosaic-org/mosaic/files/5716851/mosaic-3.log)

> After exiting neovim, there are still some leftover parts that are not cleared properly. Not sure but I think those only happen in the area where there was a spacing issue previously.

`cat`ing this log file have proper output(without the glitch seen in the gif). So this issue might be in rendering.

**Comment by imsnif** [comment]
I think this should now be fixed, @henil. Could you please check? Just merged to main

**Comment by henil** [comment]
The first part is fixed.

> After exiting neovim, there are still some leftover parts that are not cleared properly. Not sure but I think those only happen in the area where there was a spacing issue previously. Issue #93

But I can still reproduce this. I have checked the log and this does not happen when I `cat` the file in either in mosaic or outside of mosaic, so this just might be a rendering issue. But that's just a guess :)

[mosaic-3.log](https://github.com/mosaic-org/mosaic/files/5799791/mosaic-3.log)

**Comment by imsnif** [comment]
@henil - thanks for checking this. Let's try to get the log to reproduce this, otherwise it will be a little hard for me to debug.
Could you try to:
1. Delete the logfile
2. Start Mosaic
3. Reproduce the issue
4. Immediately after you reproduce the issue, close mosaic with <ctrl-g q>
5. Paste the log here

It might be that something is overriding reproduction in the log...
Also, once you do that, if you could give me the exact terminal window size using `tput cols` and `tput lines` that would be great.

**Comment by henil** [comment]
> Immediately after you reproduce the issue, close mosaic with 

I suppose you meant closing the terminal itself, so it doesn't contain any extra keys? Let me know if you meant anything else

Uploading the files:

- [log.txt](https://github.com/mosaic-org/mosaic/files/5800681/log.txt)
- [mosaic-3.log](https://github.com/mosaic-org/mosaic/files/5800682/mosaic-3.log)

> $ tput cols
> 149

> $ tput lines
> 36

**Comment by imsnif** [comment]
Ah, github deleted part of that line, sorry ;P
I meant `close mosaic with <ctrl-g q>` :)

I will check this later, but until I do: can you reproduce this when you `cat` inside mosaic? If you can't then I won't be able to either and then we need to find another way to reproduce it...

**Comment by henil** [comment]
Exited with `<ctrl-g> q`.

I thought you meant only `log.txt` file. Attaching both the files now:

- [log.txt](https://github.com/mosaic-org/mosaic/files/5800739/log.txt)
- [mosaic-3.log](https://github.com/mosaic-org/mosaic/files/5800740/mosaic-3.log)

And as I said previously, I am not able to reproduce this by `cat`ing this inside or outside of mosaic :sweat_smile: . What are some other ways we could reproduce this?

**Comment by imsnif** [comment]
> And as I said previously, I am not able to reproduce this by `cat`ing this inside or outside of mosaic sweat_smile . What are some other ways we could reproduce this?

Ah, that's too bad :/
Hrm. Maybe we can try reproducing this the "application" way, by me trying to replicate your environment. Harder, but...
Can you try to "simplify" the situation as much as possible? eg. run neovim without any plugins and reproduce this with a file I have as well?

**Comment by henil** [comment]
I had tested this with bare neovim (i.e. no plugins running with `nvim -u NONE`), and it still happens. This works on vim too.

1. Paste this function in your `config.fish` file temporarily. (it just shows disk usage)
```fish
function fish_greeting
    echo -ne (\
        df -l -h | grep -E 'dev/(xvda|sd|mapper)' | \
        awk '{printf "\\\\t%s\\\\t%4s / %4s  %s\\\\n\n", $6, $3, $2, $5}' | \
        sed -e 's/^\(.*\([8][5-9]\|[9][0-9]\)%.*\)$/\\\\e[0;31m\1\\\\e[0m/' -e 's/^\(.*\([7][5-9]\|[8][0-4]\)%.*\)$/\\\\e[0;33m\1\\\\e[0m/' | \
        paste -sd ''\
    )
    echo
end
```
2. Create a file name `init.vim` 
```bash
touch /tmp/init.vim
# Changing colorscheme just to better see it. It happens without it too.
echo "colorscheme blue" > /tmp/init.vim
```
3. Open vim/neovim with config file set as it.
> vim -u /tmp/init.vim

4. Quit vim
5. Some places still have the color left from vim window.

I have a strong feeling that this is related to #93. Issue is only at places that had spacing issues previously. 

Maybe screen renderer skips re-rendering tabs for some reason?

**Comment by imsnif** [comment]
Thanks for the detailed reproduction, @henil - I will give this a shot this afternoon and update.

**Comment by imsnif** [comment]
@henil, I'm getting a lot of:

```
E10: \ should be followed by /, ? or &                                                                              
line    1:                                                                                                          
E116: Invalid arguments for function call('printf',                                                                 
line    2:                                                                                                          
E10: \ should be followed by /, ? or &                                                                              
line    3:                                                                                                          
E10: \ should be followed by /, ? or &                                                                              
line    4:                                                                                                          
E10: \ should be followed by /, ? or &                                                                              
line    1:                                                                                                          
E116: Invalid arguments for function call('printf',                                                                 
line    2:                                                                                                          
E10: \ should be followed by /, ? or &                                                                              
line    3:                                                                                                          
E10: \ should be followed by /, ? or &                                                                              
line    4:                                                                                                          
E10: \ should be followed by /, ? or &                                                                              
Press ENTER or type command to continue                                                                             

```

When I try to start vim in this way. Do you think you can find a simpler way to reproduce this?

**Comment by imsnif** [comment]
Your trick with the `-u` did help me reproduce this though: https://github.com/mosaic-org/mosaic/issues/139 :)

**Comment by imsnif** [comment]
@henil - I merged some fixed into main that I think might have fixed this. Could you please check when you have a chance?

**Comment by henil** [comment]
@imsnif unfortunately not, I am still able to reproduce this.

**Comment by imsnif** [comment]
@henil - I want to have a go at this issue again. Since we're not able to reproduce it with debug logs, could you post a gif of what happens here (like your excellent gift for the previous evolution of this issue)? I'll try to see how to reproduce it.

**Comment by henil** [comment]
So this time I have tried to keep it as simple as possible. the `init.vim` file only contain a single line. 
> colorscheme blue

This is only to make the bug properly visible.

 And I will emphasize again, this might be related to #93 and tabs not being able to render again properly somehow. But this does not happen in all the places where there was a space issue previously so I am not sure.

https://user-images.githubusercontent.com/71698300/105461627-7febe700-5cb3-11eb-8804-5e54ee408b9c.mp4

**Comment by imsnif** [comment]
Hey @henil - the video really helped. I think it's the first time I actually understood the issue :)
I was able to reproduce it with the fixture we have of your prompt - so didn't need to configure my environment around it - which as you know is a very painful step that I'd usually rather avoid.
I fixed the issue and merged it to main. Can you please make sure it works for you?

**Comment by henil** [comment]
> I was able to reproduce it with the fixture we have of your prompt - so didn't need to configure my environment around it - which as you know is a very painful step that I'd usually rather avoid.

I know that feeling :laughing: 

>  I fixed the issue and merged it to main. Can you please make sure it works for you?

Yes, Thank you! Fixed by #152
